show field validation errors on other form tabs

agoelzer opened the pull request:
originally field validation errors could have been shown on tabs which are not selected and therefore invisible to the user when they click the "create" or "save" buttons on the form. Now it will automatically switch to the first tab having an error. Because of delayed state changes I had to keep track of the errors during the validation phase.

sivanov-nuodb commented:
@agoelzer, I suggested previously displaying red badges with validation errors on tabs so that users can see them all. If you have multiple tabs with errors, you will be navigated to the first tab and then you need to click "Create" again to find out that there are more errors.

BTW I've tested with errors on different tabs and it breaks the view:
<img width="1061" alt="Screenshot 2025-02-14 at 10 24 09" src="https://github.com/user-attachments/assets/0bb9d27f-91d5-420f-b49d-16e05b45c2ab" />

To reproduce:
1. Create a new backup policy
2. Omit the required name
3. Input invalid scope (on the second tab)
4. Click "Create" which will get you back to first tab
5. Fix the name
6. Click "Create" again

agoelzer commented:
@sivanov-nuodb I added the red badges to the tabs if any fields have errors (or missing data). Also fixed the empty tab issue.
Note that we have pre-submit validation (on the client) and server errors validation after submission. The pre-submit validation shows directly on the fields and the badges on the tab, but the server validation cannot be verified pre-submit or assigned to certain fields on the client, so those errors show at the bottom of the tabs container.

sivanov-nuodb commented:
> @sivanov-nuodb I added the red badges to the tabs if any fields have errors (or missing data). Also fixed the empty tab issue. Note that we have pre-submit validation (on the client) and server errors validation after submission.

Thanks for adding badges and fixing the issue!

> The pre-submit validation shows directly on the fields and the badges on the tab, but the server validation cannot be verified pre-submit or assigned to certain fields on the client, so those errors show at the bottom of the tabs container.

This is fine.

I played around with the live environment on `dbaas-570` and found some more issues.

* the red badge doesn't appear even if there is validation error
<img width="1038" alt="Screenshot 2025-02-18 at 16 02 40" src="https://github.com/user-attachments/assets/4f3e0d24-35d3-4d37-96cb-11eb01f80c5e" />
* the red badge appear but the validation error is not shown
<img width="1059" alt="Screenshot 2025-02-18 at 16 05 24" src="https://github.com/user-attachments/assets/515a0be9-ba24-4fe9-a857-f85913ad7d37" />

agoelzer commented:
> I played around with the live environment on `dbaas-570` and found some more issues. They might be related to the way how validation errors are rendered (on field focus/change).
I placed the red badges if any of the fields are invalid, even if not touched yet - I changed it now that it only shows if field errors are shown. Also fixed the bug where hierarchical fields (ie. `scope` field) where processed incorrectly.

sivanov-nuodb commented on ui/public/theme/base.css:
```diff
@@ -824,7 +824,8 @@ button:focus {
     display: flex;
-    flex-direction: column;
+    flex-direction: row;
+        align-items: center;
```
Nit: Strange indentation.

agoelzer replied:
fixed - sometimes the code formatter is doing those strange indentations....

sivanov-nuodb commented on ui/public/theme/base.css:
```diff
@@ -860,4 +861,14 @@ button:focus {
+.NuoBadge {
+    min-width: 10px;
+    min-height: 10px;
+    background-color: red;
```
Nit: I suggest a bit darker (e.g. `#950606`). 
Suggest making the color another CSS class. Currently, the badge is only red but you might want to create blue for other purposes.

sivanov-nuodb commented on ui/public/theme/base.css:
```diff
@@ -860,4 +861,14 @@ button:focus {
+    min-height: 10px;
+    background-color: red;
+    margin: 0 0 0 5px;
+    border-radius: 99999px;
```
Nit: Does `border-radius: 50%;` works?

agoelzer replied:
50% only works if it is guaranteed to be a circle - if the badge has a large number included (i.e. 1000), 50% would make an ellipsis instead of rounding the corners (and expanding horizontally).

sivanov-nuodb commented on ui/src/components/controls/Tabs.tsx:
```diff
@@ -14,11 +14,17 @@ export function Tab({ children }: TabProps) {
 
-export function Tabs({ children }: TabsProps) {
+function badge(count: number) {
+    return <div className="NuoBadge">{count < 0 ? "" : count}</div>
```
Having number of errors is a nice touch.

sivanov-nuodb commented on ui/src/components/pages/parts/CreateEditEntry.tsx:
```diff
@@ -267,26 +272,83 @@ function CreateEditEntry({ schema, path, data, readonly, org, t }: TempAny) {
+        return -1;
+    }
+
+    function getTabIndexesWithErrors(): number[] {
```
Counting the number of errors in each tab and displaying it in the badge would be even better.
I saw that you have support for that in the `Tabs` component.

agoelzer replied:
I tried including the number but then the red dot becomes too big to make the numbers fit, so I opted without the number.

sivanov-nuodb reviewed (approved):
Looks good.
I left some minor suggestions.
